src/bootstrap/channel.rs

```diff
@@ -18,19 +19,40 @@ pub enum GitInfo {
     /// If the info should be used (`ignore_git` is false), this will be
     /// `Some`, otherwise it will be `None`.
     Present(Option<Info>),
+    /// This is not a git repostory, but the info can be fetched from the
+    /// `git-commit-info` file.
+    RecordedForTarball(Info),
 }
 
 pub struct Info {
-    commit_date: String,
-    sha: String,
-    short_sha: String,
+    pub commit_date: String,
+    pub sha: String,
+    pub short_sha: String,
 }
 
 impl GitInfo {
     pub fn new(ignore_git: bool, dir: &Path) -> GitInfo {
         // See if this even begins to look like a git dir
         if !dir.join(".git").exists() {
-            return GitInfo::Absent;
+            if let Ok(info) = fs::read_to_string(dir.join("git-commit-info")) {
+                let mut lines = info.lines();
+                let sha = lines.next();
+                let short_sha = lines.next();
+                let commit_date = lines.next();
+
+                // If the file is malformed, just don't use it at all.
+                if sha.is_none() || short_sha.is_none() || commit_date.is_none() {
+                    return GitInfo::Absent;
+                }
+
+                return GitInfo::RecordedForTarball(Info {
+                    commit_date: commit_date.unwrap().to_owned(),
+                    sha: sha.unwrap().to_owned(),
+                    short_sha: short_sha.unwrap().to_owned(),
+                });
```

Comment:
nit: you can avoid unwrap here
```suggestion
                let info = match (commit_date, sha, short_sha) {
                    Some((date, sha, short_sha)) => Info { ... },
                    _ => return GitInfo::Absent, // If the file is malformed, just don't use it at all
               };
                return GitInfo::RecordedForTarball(info);
```